lib/puppet/functions/binary_file.rb

```diff
@@ -0,0 +1,25 @@
+# Loads a binary file from a module or file system and returns its contents as a Binary.
+# (Documented in 3.x stub)
+#
+# @since 4.8.0
+#
+Puppet::Functions.create_function(:binary_file, Puppet::Functions::InternalFunction) do
+  dispatch :binary_file do
+    scope_param
+    param 'String', :path
+  end
+
+  def binary_file(scope, unresolved_path)
+    path = nil
+    found = Puppet::Parser::Files.find_file(unresolved_path, scope.compiler.environment)
+    if found && Puppet::FileSystem.exist?(found)
+      path = found
+    end
+
+    if path
+      Puppet::Pops::Types::PBinaryType::Binary.from_binary_string(Puppet::FileSystem.binread(path))
```

Comment:
Sure. Given though that, there is an issue currently with ENOENT vs. permission failure on JRuby and that the expected frequency of calls to `binary_file` is both low, and minor in relation to actually reading the file, the optimization of this would be best performed later - and then looking at file reading and existence checks in general.